File: android-core/src/test/java/com/mparticle/identity/MParticleUserTest.java
package com.mparticle.identity;

import com.mparticle.MParticle;
import com.mparticle.MockMParticle;
import com.mparticle.UserAttributeListener;
import com.mparticle.consent.ConsentState;

import junit.framework.Assert;

import org.json.JSONArray;
import org.json.JSONObject;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Mockito;

import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNotSame;
import static org.junit.Assert.assertTrue;
import static org.mockito.Matchers.eq;

public class MParticleUserTest {
    private MParticle mp;
    private IdentityApi id;
    private Long defaultMpId = 1L;

    @Before
    public void setup() {
        MParticle.setInstance(new MockMParticle());
        mp = MParticle.getInstance();
        id = mp.Identity();
        Mockito.when(mp.Identity().mConfigManager.getMpid()).thenReturn(defaultMpId);

    }

    /**
     * this tests that when you add a call setUserIdentity, with a user name and id that already exists,
     * we will not log it
     */
    @Test
    public void testAddExistingUserIdentity() throws Exception {
        JSONArray identities = new JSONArray();
        identities.put(new JSONObject("{ \"n\": 8, \"i\": \"alias test\", \"dfs\": 1473869816521, \"f\": true }"));
        Mockito.when(id.mMessageManager.getUserIdentityJson(defaultMpId)).thenReturn(identities);
        Mockito.when(id.mMessageManager.logUserIdentityChangeMessage(Mockito.any(JSONObject.class), Mockito.any(JSONObject.class), Mockito.any(JSONArray.class), eq(defaultMpId))).thenThrow(new AssertionError("Should not log redundent User Identity"));
        Map<MParticle.IdentityType, String> ids = new HashMap<MParticle.IdentityType, String >();
        ids.put(MParticle.IdentityType.Alias, "alias test");
        ((MParticleUserImpl)id.getCurrentUser()).setUserIdentities(ids);
    }

    @Test
    public void testChangeUserIdentity() throws Exception {
        JSONArray identities = new JSONArray();
        identities.put(new JSONObject("{ \"n\": 7, \"i\": \"email value 1\", \"dfs\": 1473869816521, \"f\": true }"));
        Mockito.when(id.mMessageManager.getUserIdentityJson(1)).thenReturn(identities);
        ((MParticleUserImpl)id.getCurrentUser()).setUserIdentity(MParticle.IdentityType.Email, "email value 2");
        ArgumentCaptor<JSONObject> argument1 = ArgumentCaptor.forClass(JSONObject.class);
        ArgumentCaptor<JSONObject> argument2 = ArgumentCaptor.forClass(JSONObject.class);
        ArgumentCaptor<JSONArray> argument3 = ArgumentCaptor.forClass(JSONArray.class);
        Mockito.verify(id.mMessageManager,
                Mockito.times(1)).
                logUserIdentityChangeMessage(
                        argument1.capture(),
                        argument2.capture(),
                        argument3.capture(),
                        eq(defaultMpId)
                );
        JSONObject oldIdentity = argument2.getValue();
        assertEquals(oldIdentity.get("i"), "email value 1");
        assertEquals(oldIdentity.get("n"), 7);
        assertEquals(oldIdentity.getDouble("dfs"), 1473869816521d, 100);
        assertEquals(oldIdentity.get("f"), true);
        JSONObject newIdentity = argument1.getValue();
        assertEquals(newIdentity.get("i"), "email value 2");
        assertEquals(newIdentity.get("n"), 7);
        assertEquals(newIdentity.getDouble("dfs"), 1473869816521d, 100);
        assertEquals(newIdentity.get("f"), false);
        Mockito.verify(MParticle.getInstance().Internal().getKitManager(), Mockito.times(1)).setUserIdentity(eq("email value 2"), eq(MParticle.IdentityType.Email));
        JSONArray allIdentities = argument3.getValue();
        assertEquals(1, allIdentities.length());
    }


    @Test
    public void testRemoveUserAttribute() throws Exception {
        IdentityApi mp = MParticle.getInstance().Identity();
        assertFalse(mp.getCurrentUser().removeUserAttribute(null));
        assertFalse(mp.getCurrentUser().removeUserAttribute(""));
        assertTrue(mp.getCurrentUser().removeUserAttribute("test"));
        Mockito.verify(mp.mMessageManager, Mockito.times(1)).removeUserAttribute("test", 1);
        Mockito.verify(MParticle.getInstance().Internal().getKitManager(), Mockito.times(1)).removeUserAttribute("test", 1);
    }

    @Test
    public void testRemoveUserTag() throws Exception {
        assertFalse(id.getCurrentUser().removeUserAttribute(null));
        assertFalse(id.getCurrentUser().removeUserAttribute(""));
        assertTrue(id.getCurrentUser().removeUserAttribute("test"));
        Mockito.verify(id.mMessageManager, Mockito.times(1)).removeUserAttribute("test", 1);
        Mockito.verify(mp.Internal().getKitManager(), Mockito.times(1)).removeUserAttribute("test", 1);
    }


    @Test
    public void testGetAllUserAttributes1() throws Exception {
        UserAttributeListener listener = Mockito.mock(UserAttributeListener.class);
        id.getCurrentUser().getUserAttributes(listener);
        Mockito.verify(mp.Identity().mMessageManager, Mockito.times(1)).getUserAttributes(listener, defaultMpId);
    }


    @Test
    public void testRemoveUserIdentityWhenNoneExist() throws Exception {
        JSONArray identities = new JSONArray();
        Mockito.when(mp.Identity().mMessageManager.getUserIdentityJson(defaultMpId)).thenReturn(identities);
        ((MParticleUserImpl)id.getCurrentUser()).setUserIdentity(MParticle.IdentityType.Alias, null);
        Mockito.verify(mp.Identity().mMessageManager,
                Mockito.times(0)).
                logUserIdentityChangeMessage(
                        Mockito.any(JSONObject.class),
                        Mockito.any(JSONObject.class),
                        Mockito.any(JSONArray.class),
                        eq(defaultMpId)
                );
        Mockito.verify(mp.Internal().getKitManager(), Mockito.times(0)).removeUserIdentity(Mockito.any(MParticle.IdentityType.class));
    }

    @Test
    public void testRemoveUserIdentity() throws Exception {
        JSONArray identities = new JSONArray();
        identities.put(new JSONObject("{ \"n\": 7, \"i\": \"email value 1\", \"dfs\": 1473869816521, \"f\": true }"));
        Mockito.when(mp.Identity().mMessageManager.getUserIdentityJson(defaultMpId)).thenReturn(identities);
        ((MParticleUserImpl)id.getCurrentUser()).setUserIdentity(MParticle.IdentityType.Email, null);
        ArgumentCaptor<JSONObject> argument2 = ArgumentCaptor.forClass(JSONObject.class);
        ArgumentCaptor<JSONArray> argument3 = ArgumentCaptor.forClass(JSONArray.class);
        ArgumentCaptor<Long> argument4 = ArgumentCaptor.forClass(Long.class);
        Mockito.verify(mp.Identity().mMessageManager,
                Mockito.times(1)).
                logUserIdentityChangeMessage(
                        Mockito.isNull(JSONObject.class),
                        argument2.capture(),
                        argument3.capture(),
                        argument4.capture()
                );
        JSONObject oldIdentity = argument2.getValue();
        assertEquals(oldIdentity.get("i"), "email value 1");
        assertEquals(oldIdentity.get("n"), 7);
        assertEquals(oldIdentity.getDouble("dfs"), 1473869816521d, 100);
        assertEquals(oldIdentity.get("f"), true);
        assertTrue(argument4.getValue().equals(defaultMpId));
        Mockito.verify(mp.Internal().getKitManager(), Mockito.times(1)).removeUserIdentity(MParticle.IdentityType.Email);
        JSONArray allIdentities = argument3.getValue();
        assertEquals(0, allIdentities.length());
    }

    @Test
    public void testAddUserIdentity() throws Exception {
        JSONArray identities = new JSONArray();
        identities.put(new JSONObject("{ \"n\": 7, \"i\": \"email value 1\", \"dfs\": 1473869816521, \"f\": true }"));
        Mockito.when(mp.Identity().mMessageManager.getUserIdentityJson(Mockito.anyInt())).thenReturn(identities);
        ((MParticleUserImpl)id.getCurrentUser()).setUserIdentity(MParticle.IdentityType.Alias, "alias test");
        ArgumentCaptor<JSONObject> argument2 = ArgumentCaptor.forClass(JSONObject.class);
        ArgumentCaptor<JSONArray> argument3 = ArgumentCaptor.forClass(JSONArray.class);
        ArgumentCaptor<Long> argument4 = ArgumentCaptor.forClass(Long.class);
        Mockito.verify(mp.Identity().mMessageManager,
                Mockito.times(1)).
                logUserIdentityChangeMessage(
                        argument2.capture(),
                        Mockito.isNull(JSONObject.class),
                        argument3.capture(),
                        argument4.capture()
                );
        JSONObject oldIdentity = argument2.getValue();
        assertEquals(oldIdentity.get("i"), "alias test");
        assertEquals(oldIdentity.get("n"), MParticle.IdentityType.Alias.getValue());
        assertEquals(oldIdentity.getDouble("dfs"), System.currentTimeMillis(), 1000);
        assertEquals(oldIdentity.get("f"), true);
        assertTrue(argument4.getValue().equals(defaultMpId));
        Mockito.verify(mp.Internal().getKitManager(), Mockito.times(1)).setUserIdentity(eq("alias test"), eq(MParticle.IdentityType.Alias));

        JSONArray allIdentities = argument3.getValue();
        assertEquals(2, allIdentities.length());
    }
}

File: android-core/src/test/kotlin/com/mparticle/identity/MParticleUserTest.kt
package com.mparticle.identity

import com.mparticle.MParticle
import com.mparticle.MParticle.IdentityType
import com.mparticle.MockMParticle
import com.mparticle.TypedUserAttributeListener
import com.mparticle.UserAttributeListener
import org.json.JSONArray
import org.json.JSONObject
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers
import org.mockito.Mockito

class MParticleUserTest {
    private var mp: MParticle? = null
    private var id: IdentityApi? = null
    private val defaultMpId = 1L

    @Before
    fun setup() {
        MParticle.setInstance(MockMParticle())
        mp = MParticle.getInstance()
        id = mp?.Identity()
        Mockito.`when`(mp?.Identity()?.mConfigManager?.mpid).thenReturn(defaultMpId)
    }

    /**
     * This tests that when you add a call setUserIdentity, with a user name and id that already exists,
     * we will not log it.
     */
    @Test
    @Throws(Exception::class)
    fun testAddExistingUserIdentity() {
        val identities = JSONArray()
        identities.put(JSONObject("{ \"n\": 8, \"i\": \"alias test\", \"dfs\": 1473869816521, \"f\": true }"))
        Mockito.`when`(id?.mMessageManager?.getUserIdentityJson(defaultMpId)).thenReturn(identities)
        Mockito.`when`(
            id?.mMessageManager?.logUserIdentityChangeMessage(
                Mockito.any(
                    JSONObject::class.java
                ),
                Mockito.any(JSONObject::class.java),
                Mockito.any(
                    JSONArray::class.java
                ),
                ArgumentMatchers.eq<Long>(defaultMpId)
            )
        ).thenThrow(
            AssertionError("Should not log redundent User Identity")
        )
        val ids = HashMap<IdentityType, String>()
        ids[IdentityType.Alias] = "alias test"
        (id?.currentUser as MParticleUserImpl?)?.userIdentities = ids
    }

    @Test
    @Throws(Exception::class)
    fun testChangeUserIdentity() {
        val identities = JSONArray()
        identities.put(JSONObject("{ \"n\": 7, \"i\": \"email value 1\", \"dfs\": 1473869816521, \"f\": true }"))
        Mockito.`when`(id?.mMessageManager?.getUserIdentityJson(1)).thenReturn(identities)
        (id?.currentUser as MParticleUserImpl?)?.setUserIdentity(
            IdentityType.Email,
            "email value 2"
        )
        val argument1 = ArgumentCaptor.forClass(
            JSONObject::class.java
        )
        val argument2 = ArgumentCaptor.forClass(
            JSONObject::class.java
        )
        val argument3 = ArgumentCaptor.forClass(
            JSONArray::class.java
        )
        Mockito.verify(
            id!!.mMessageManager,
            Mockito.times(1)
        ).logUserIdentityChangeMessage(
            argument1.capture(),
            argument2.capture(),
            argument3.capture(),
            ArgumentMatchers.eq<Long>(defaultMpId)
        )
        val oldIdentity = argument2.value
        Assert.assertEquals(oldIdentity["i"], "email value 1")
        Assert.assertEquals(oldIdentity["n"], 7)
        Assert.assertEquals(oldIdentity.getDouble("dfs"), 1473869816521.0, 100.0)
        Assert.assertEquals(oldIdentity["f"], true)
        val newIdentity = argument1.value
        Assert.assertEquals(newIdentity["i"], "email value 2")
        Assert.assertEquals(newIdentity["n"], 7)
        Assert.assertEquals(newIdentity.getDouble("dfs"), 1473869816521.0, 100.0)
        Assert.assertEquals(newIdentity["f"], false)
        Mockito.verify(
            MParticle.getInstance()?.Internal()?.kitManager,
            Mockito.times(1)
        )
            ?.setUserIdentity(
                ArgumentMatchers.eq("email value 2"),
                ArgumentMatchers.eq(
                    IdentityType.Email
                )
            )
        val allIdentities = argument3.value
        Assert.assertEquals(1, allIdentities.length().toLong())
    }

    @Test
    @Throws(Exception::class)
    fun testRemoveUserAttribute() {
        val mp = MParticle.getInstance()?.Identity()
        mp?.currentUser?.removeUserAttribute("")?.let { Assert.assertFalse(it) }
        mp?.currentUser?.removeUserAttribute("")?.let { Assert.assertFalse(it) }
        mp?.currentUser?.removeUserAttribute("test")?.let { Assert.assertTrue(it) }
        Mockito.verify(
            mp?.mMessageManager,
            Mockito.times(1)
        )?.removeUserAttribute("test", 1)
        Mockito.verify(MParticle.getInstance()!!.Internal().kitManager, Mockito.times(1))
            .removeUserAttribute("test", 1)
    }

    @Test
    @Throws(Exception::class)
    fun testRemoveUserTag() {
        id?.currentUser?.removeUserAttribute("")?.let { Assert.assertFalse(it) }
        id?.currentUser?.removeUserAttribute("")?.let { Assert.assertFalse(it) }
        id?.currentUser?.removeUserAttribute("test")?.let { Assert.assertTrue(it) }
        Mockito.verify(
            id?.mMessageManager,
            Mockito.times(1)
        )?.removeUserAttribute("test", 1)
        Mockito.verify(
            mp?.Internal()?.kitManager,
            Mockito.times(1)
        )?.removeUserAttribute("test", 1)
    }

    @Test
    @Throws(Exception::class)
    fun testGetAllUserAttributes() {
        val listener = Mockito.mock(
            UserAttributeListener::class.java
        )
        id?.currentUser?.getUserAttributes(listener)
        Mockito.verify(
            mp?.Identity()?.mMessageManager,
            Mockito.times(1)
        )?.getUserAttributes(
            ArgumentMatchers.any(
                UserAttributeListenerWrapper::class.java
            ),
            ArgumentMatchers.eq<Long>(defaultMpId)
        )
        val typedListener = Mockito.mock(
            TypedUserAttributeListener::class.java
        )
        id?.currentUser?.getUserAttributes(typedListener)
        Mockito.verify(
            mp?.Identity()?.mMessageManager,
            Mockito.times(2)
        )?.getUserAttributes(
            ArgumentMatchers.any(
                UserAttributeListenerWrapper::class.java
            ),
            ArgumentMatchers.eq<Long>(defaultMpId)
        )
    }

    @Test
    @Throws(Exception::class)
    fun testRemoveUserIdentityWhenNoneExist() {
        val identities = JSONArray()
        Mockito.`when`(mp?.Identity()?.mMessageManager?.getUserIdentityJson(defaultMpId))
            .thenReturn(identities)
        (id?.currentUser as MParticleUserImpl?)?.setUserIdentity(IdentityType.Alias, null)
        Mockito.verify(
            mp?.Identity()?.mMessageManager,
            Mockito.times(0)
        )?.logUserIdentityChangeMessage(
            Mockito.any(JSONObject::class.java),
            Mockito.any(JSONObject::class.java),
            Mockito.any(JSONArray::class.java),
            ArgumentMatchers.eq<Long>(defaultMpId)
        )
        Mockito.verify(
            mp?.Internal()?.kitManager,
            Mockito.times(0)
        )?.removeUserIdentity(
            Mockito.any(
                IdentityType::class.java
            )
        )
    }

    @Test
    @Throws(Exception::class)
    fun testRemoveUserIdentity() {
        val identities = JSONArray()
        identities.put(JSONObject("{ \"n\": 7, \"i\": \"email value 1\", \"dfs\": 1473869816521, \"f\": true }"))
        Mockito.`when`(mp?.Identity()?.mMessageManager?.getUserIdentityJson(defaultMpId))
            .thenReturn(identities)
        (id?.currentUser as MParticleUserImpl?)?.setUserIdentity(IdentityType.Email, null)
        val argument2 = ArgumentCaptor.forClass(
            JSONObject::class.java
        )
        val argument3 = ArgumentCaptor.forClass(
            JSONArray::class.java
        )
        val argument4 = ArgumentCaptor.forClass(
            Long::class.java
        )
        Mockito.verify(
            mp?.Identity()?.mMessageManager,
            Mockito.times(1)
        )?.logUserIdentityChangeMessage(
            Mockito.isNull(JSONObject::class.java),
            argument2.capture(),
            argument3.capture(),
            argument4.capture()
        )
        val oldIdentity = argument2.value
        Assert.assertEquals(oldIdentity["i"], "email value 1")
        Assert.assertEquals(oldIdentity["n"], 7)
        Assert.assertEquals(oldIdentity.getDouble("dfs"), 1473869816521.0, 100.0)
        Assert.assertEquals(oldIdentity["f"], true)
        Assert.assertTrue(argument4.value == defaultMpId)
        Mockito.verify(
            mp?.Internal()?.kitManager,
            Mockito.times(1)
        )?.removeUserIdentity(
            IdentityType.Email
        )
        val allIdentities = argument3.value
        Assert.assertEquals(0, allIdentities.length().toLong())
    }

    @Test
    @Throws(Exception::class)
    fun testAddUserIdentity() {
        val identities = JSONArray()
        identities.put(JSONObject("{ \"n\": 7, \"i\": \"email value 1\", \"dfs\": 1473869816521, \"f\": true }"))
        Mockito.`when`(mp?.Identity()?.mMessageManager?.getUserIdentityJson(Mockito.anyLong()))
            .thenReturn(identities)
        (id?.currentUser as MParticleUserImpl?)?.setUserIdentity(IdentityType.Alias, "alias test")
        val argument2 = ArgumentCaptor.forClass(
            JSONObject::class.java
        )
        val argument3 = ArgumentCaptor.forClass(
            JSONArray::class.java
        )
        val argument4 = ArgumentCaptor.forClass(
            Long::class.java
        )
        Mockito.verify(
            mp?.Identity()?.mMessageManager,
            Mockito.times(1)
        )?.logUserIdentityChangeMessage(
            argument2.capture(),
            Mockito.isNull(JSONObject::class.java),
            argument3.capture(),
            argument4.capture()
        )
        val oldIdentity = argument2.value
        Assert.assertEquals(oldIdentity["i"], "alias test")
        Assert.assertEquals(
            oldIdentity["n"],
            IdentityType.Alias.value
        )
        Assert.assertEquals(
            oldIdentity.getDouble("dfs"),
            System.currentTimeMillis().toDouble(),
            1000.0
        )
        Assert.assertEquals(oldIdentity["f"], true)
        Assert.assertTrue(argument4.value == defaultMpId)
        Mockito.verify(mp?.Internal()?.kitManager, Mockito.times(1))?.setUserIdentity(
            ArgumentMatchers.eq("alias test"),
            ArgumentMatchers.eq(
                IdentityType.Alias
            )
        )
        val allIdentities = argument3.value
        Assert.assertEquals(2, allIdentities.length().toLong())
    }
}
